<!DOCTYPE html><html class=writer-html5 lang=en> <head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width, initial-scale=1.0"><meta name=description content="ChIP-seq pipeline"><meta name=author content="CCR Collaborative Bioinformatics Resource"><link rel="shortcut icon" href=img/favicon.ico><title>CHAMPAGNE</title><link rel=stylesheet href=css/theme.css><link rel=stylesheet href=css/theme_extra.css><link rel=stylesheet href=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css><link href=css/custom.css rel=stylesheet><link href=css/version-select.css rel=stylesheet><script>
        // Current page data
        var mkdocs_page_name = "Home";
        var mkdocs_page_input_path = "index.md";
        var mkdocs_page_url = null;
      </script><!--[if lt IE 9]>
      <script src="js/html5shiv.min.js"></script>
    <![endif]--><script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js></script><script>hljs.highlightAll();</script></head> <body class=wy-body-for-nav role=document> <div class=wy-grid-for-nav> <nav data-toggle=wy-nav-shift class="wy-nav-side stickynav"> <div class=wy-side-scroll> <div class=wy-side-nav-search> <a href=. class="icon icon-home"> CHAMPAGNE </a><div role=search> <form id=rtd-search-form class=wy-form action=./search.html method=get> <input type=text name=q placeholder="Search docs" aria-label="Search docs" title="Type search term here"> </form> </div> </div> <div class="wy-menu wy-menu-vertical" data-spy=affix role=navigation aria-label="Navigation menu"> <ul class=current> <li class="toctree-l1 current"><a class="reference internal current" href=#>Home</a> <ul class=current> <li class=toctree-l2><a class="reference internal" href=#set-up>Set up</a> <ul> <li class=toctree-l3><a class="reference internal" href=#biowulf>Biowulf</a> </li> </ul> </li> <li class=toctree-l2><a class="reference internal" href=#usage>Usage</a> </li> <li class=toctree-l2><a class="reference internal" href=#help-contributing>Help &amp; Contributing</a> </li> <li class=toctree-l2><a class="reference internal" href=#references>References</a> </li> </ul> </li> </ul> <p class=caption><span class=caption-text>Usage</span></p> <ul> <li class=toctree-l1><a href=getting-started/ class="reference internal">Getting Started</a> </li> <li class=toctree-l1><a href=manifests/ class="reference internal">Creating Manifests</a> </li> <li class=toctree-l1><a href=workflow/ class="reference internal">Workflow Overview</a> </li> </ul> <p class=caption><span class=caption-text>Developer Guide</span></p> <ul> <li class=toctree-l1><a href=contributing/ class="reference internal">How to contribute</a> </li> <li class=toctree-l1><a href=release-guide/ class="reference internal">Release guide</a> </li> </ul> <ul> <li class=toctree-l1><a href=contributors/ class="reference internal">Contributors</a> </li> </ul> <ul> <li class=toctree-l1><a href=changelog/ class="reference internal">Changelog</a> </li> </ul> </div> </div> </nav> <section data-toggle=wy-nav-shift class=wy-nav-content-wrap> <nav class=wy-nav-top role=navigation aria-label="Mobile navigation menu"> <i data-toggle=wy-nav-top class="fa fa-bars"></i> <a href=.>CHAMPAGNE</a> </nav> <div class=wy-nav-content> <div class=rst-content><div role=navigation aria-label="breadcrumbs navigation"> <ul class=wy-breadcrumbs> <li><a href=. class="icon icon-home" aria-label=Docs></a></li> <li class="breadcrumb-item active">Home</li> <li class=wy-breadcrumbs-aside> <a href=https://github.com/CCBR/CHAMPAGNE/edit/main/index.md>Edit on CCBR/CHAMPAGNE</a> </li> </ul> <hr> </div> <div role=main class=document itemscope=itemscope itemtype=http://schema.org/Article> <div class=section itemprop=articleBody> <!-- the README is inserted here automatically--> <h1 id=champagne>CHAMPAGNE üçæ<a class=headerlink href=#champagne title="Permanent link">&para;</a></h1> <p><strong>CH</strong>rom<strong>A</strong>tin i<strong>M</strong>muno <strong>P</strong>recipit<strong>A</strong>tion sequencin<strong>G</strong> a<strong>N</strong>alysis pip<strong>E</strong>line</p> <p><a href=https://github.com/CCBR/CHAMPAGNE/actions/workflows/build.yml><img alt=build src=https://github.com/CCBR/CHAMPAGNE/actions/workflows/build.yml/badge.svg></a> <a href=https://ccbr.github.io/CHAMPAGNE><img alt=docs src=https://github.com/CCBR/CHAMPAGNE/actions/workflows/docs-mkdocs.yml/badge.svg></a> <a href=https://doi.org/10.5281/zenodo.10516079><img alt=DOI src=https://zenodo.org/badge/DOI/10.5281/zenodo.10516079.svg></a> <a href=https://github.com/CCBR/CHAMPAGNE/releases/latest><img alt=release src="https://img.shields.io/github/v/release/CCBR/CHAMPAGNE?color=blue&label=latest%20release"></a></p> <p>üöß <strong>This project is under active development. It is not yet ready for production use.</strong> üöß</p> <h2 id=set-up>Set up<a class=headerlink href=#set-up title="Permanent link">&para;</a></h2> <p>Champagne is installed on the <a href=#biowulf>Biowulf HPC</a>. For installation in other execution environments, refer to the <a href=https://ccbr.github.io/CHAMPAGNE/ >docs</a>.</p> <h3 id=biowulf>Biowulf<a class=headerlink href=#biowulf title="Permanent link">&para;</a></h3> <p>Champagne is available on <a href=https://hpc.nih.gov/ >Biowulf</a> in the <code>ccbrpipeliner</code> module. You'll first need to start an interactive session and create a directory from where you'll run champagne.</p> <div class=highlight><pre><span></span><code><span class=c1># start an interactive node</span>
sinteractive<span class=w> </span>--mem<span class=o>=</span>2g<span class=w> </span>--cpus-per-task<span class=o>=</span><span class=m>2</span><span class=w> </span>--gres<span class=o>=</span>lscratch:200
<span class=c1># make a working directory for your project and go to it</span>
mkdir<span class=w> </span>-p<span class=w> </span>/data/<span class=nv>$USER</span>/chipseq
<span class=nb>cd</span><span class=w> </span>/data/<span class=nv>$USER</span>/chipseq
<span class=c1># load the ccbrpipeliener module</span>
module<span class=w> </span>load<span class=w> </span>ccbrpipeliner
</code></pre></div> <h2 id=usage>Usage<a class=headerlink href=#usage title="Permanent link">&para;</a></h2> <p>Initialize and run champagne with test data:</p> <div class=highlight><pre><span></span><code><span class=c1># copy the champagne config files to your project directory.</span>
<span class=c1># --output is optional and defaults to your current working directory.</span>
champagne<span class=w> </span>init<span class=w> </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/champagne_project
<span class=c1># preview the champagne jobs that will run with the test dataset</span>
champagne<span class=w> </span>run<span class=w> </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/champagne_project<span class=w> </span>--mode<span class=w> </span><span class=nb>local</span><span class=w> </span>-profile<span class=w> </span><span class=nb>test</span><span class=w> </span>-preview
<span class=c1># launch a champagne run on slurm with the test dataset</span>
champagne<span class=w> </span>run<span class=w> </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/champagne_project<span class=w> </span>--mode<span class=w> </span>slurm<span class=w> </span>-profile<span class=w> </span>test,biowulf
</code></pre></div> <p>To run champagne on your own data, you'll need to create a sample sheet. Take a look at these examples:</p> <ul> <li><a href=/assets/samplesheet_test.csv>assets/samplesheet_test.csv</a> - mix of single and paired end reads downloaded from github.</li> <li><a href=/assets/samplesheet_test.csv>assets/samplesheet_mm10.csv</a> - single end reads on biowulf.</li> </ul> <p>Once you've created a samplesheet with paths to your fastq files, run champagne with the <code>--input</code> option to specify the path to your sample sheet:</p> <div class=highlight><pre><span></span><code>champagne<span class=w> </span>run<span class=w> </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/champagne_project<span class=w> </span>--mode<span class=w> </span>slurm<span class=w> </span>-profile<span class=w> </span>biowulf<span class=w> </span>--input<span class=w> </span>samplesheet.csv<span class=w> </span>--genome<span class=w> </span>hg38
</code></pre></div> <p>We currently support the hg38 and mm10 genomes. If you'd like to use a custom genome, you'll need the following files:</p> <ul> <li>genome fasta</li> <li>genome GTF</li> <li>blacklist fasta</li> </ul> <p>Prepare your custom reference genome with:</p> <div class=highlight><pre><span></span><code>champagne<span class=w> </span>run<span class=w> </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/champagne_project<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--mode<span class=w> </span>slurm<span class=w> </span>-profile<span class=w> </span>biowulf<span class=w> </span><span class=se>\</span>
<span class=w>    </span>-entry<span class=w> </span>MAKE_REFERENCE<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--outdir<span class=w> </span>custom_genome<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--genome<span class=w> </span>custom_genome<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--genome_fasta<span class=w> </span>genome.fasta<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--genes_gtf<span class=w> </span>genome.gtf<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--blacklist<span class=w> </span>blacklist.fasta
</code></pre></div> <p>The reference files and a config file for the genome will be written in <code>custom_genome/genome</code>.</p> <p>Then you can run champagne using your custom genome:</p> <div class=highlight><pre><span></span><code>champagne<span class=w> </span>run<span class=w> </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/champagne_project<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--mode<span class=w> </span>slurm<span class=w> </span>-profile<span class=w> </span>biowulf<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--input<span class=w> </span>samplesheet.csv<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--genome<span class=w> </span>custom_genome<span class=w> </span><span class=se>\</span>
<span class=w>    </span>-c<span class=w> </span>custom_genome/genome/custom_genome.config
</code></pre></div> <h2 id=help-contributing>Help &amp; Contributing<a class=headerlink href=#help-contributing title="Permanent link">&para;</a></h2> <p>Come across a <strong>bug</strong>? Open an <a href=https://github.com/CCBR/CHAMPAGNE/issues>issue</a> and include a minimal reproducible example.</p> <p>Have a <strong>question</strong>? Ask it in <a href=https://github.com/CCBR/CHAMPAGNE/discussions>discussions</a>.</p> <p>Want to <strong>contribute</strong> to this project? Check out the <a href=.github/CONTRIBUTING.md>contributing guidelines</a>.</p> <p><strong>General Inquiries and Collaboration:</strong> Please contact the CCBR Pipeliner team at <a href=mailto:CCBR_Pipeliner@mail.nih.gov>CCBR_Pipeliner@mail.nih.gov</a>.</p> <h2 id=references>References<a class=headerlink href=#references title="Permanent link">&para;</a></h2> <p>This repo was originally generated from the <a href=https://github.com/CCBR/CCBR_NextflowTemplate>CCBR Nextflow Template</a>. The template takes inspiration from nektool<sup id=fnref:1><a class=footnote-ref href=#fn:1>1</a></sup> and the nf-core template. If you plan to contribute your pipeline to nf-core, don't use this template -- instead follow nf-core's instructions<sup id=fnref:2><a class=footnote-ref href=#fn:2>2</a></sup>.</p> <!--- write additional info that you only want to appear in the docs index here --> <p>Information on who the pipeline was developed for, and a statement if it's only been tested on Biowulf. For example:</p> <p>It has been developed and tested solely on NIH <a href=https://hpc.nih.gov/ >HPC Biowulf</a>.</p> <p>Also include a workflow image to summarize the pipeline.</p> <div class=footnote> <hr> <ol> <li id=fn:1> <p>nektool https://github.com/beardymcjohnface/nektool&#160;<a class=footnote-backref href=#fnref:1 title="Jump back to footnote 1 in the text">&#8617;</a></p> </li> <li id=fn:2> <p>instructions for nf-core pipelines https://nf-co.re/docs/contributing/tutorials/creating_with_nf_core&#160;<a class=footnote-backref href=#fnref:2 title="Jump back to footnote 2 in the text">&#8617;</a></p> </li> </ol> </div> </div> </div><script async src="https://www.googletagmanager.com/gtag/js?id=G-D3SL9V30KL"></script> <script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-D3SL9V30KL'); // Replace with your tracking ID
</script> </div> </div> </section> </div> <div class=rst-versions role=note aria-label=Versions> <span class=rst-current-version data-toggle=rst-current-version> <span> <a href=https://github.com/CCBR/CHAMPAGNE class="fa fa-code-fork" style="color: #fcfcfc"> CCBR/CHAMPAGNE</a> </span> <span><a href=getting-started/ style="color: #fcfcfc">Next &raquo;</a></span> </span> </div> <script src=js/jquery-3.6.0.min.js></script> <script>var base_url = ".";</script> <script src=js/theme_extra.js></script> <script src=js/theme.js></script> <script src=js/custom.js></script> <script src=search/main.js></script> <script src=js/version-select.js></script> <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script> </body> </html> <!--
MkDocs version : 1.6.1
Build Date UTC : 2025-05-31 11:48:06.894495+00:00
--> 