<!DOCTYPE html><html class=writer-html5 lang=en> <head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width, initial-scale=1.0"><meta name=author content="CCR Collaborative Bioinformatics Resource"><link rel="shortcut icon" href=../../img/favicon.ico><title>Parameters - CHAMPAGNE</title><link rel=stylesheet href=../../css/theme.css><link rel=stylesheet href=../../css/theme_extra.css><link rel=stylesheet href=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css><link href=../../css/custom.css rel=stylesheet><link href=../../css/version-select.css rel=stylesheet><script>
        // Current page data
        var mkdocs_page_name = "Parameters";
        var mkdocs_page_input_path = "guide/params.md";
        var mkdocs_page_url = null;
      </script><!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]--><script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js></script><script>hljs.highlightAll();</script></head> <body class=wy-body-for-nav role=document> <div class=wy-grid-for-nav> <nav data-toggle=wy-nav-shift class="wy-nav-side stickynav"> <div class=wy-side-scroll> <div class=wy-side-nav-search> <a href=../.. class="icon icon-home"> CHAMPAGNE </a><div role=search> <form id=rtd-search-form class=wy-form action=../../search.html method=get> <input type=text name=q placeholder="Search docs" aria-label="Search docs" title="Type search term here"> </form> </div> </div> <div class="wy-menu wy-menu-vertical" data-spy=affix role=navigation aria-label="Navigation menu"> <ul> <li class=toctree-l1><a class="reference internal" href=../..>Home</a> </li> </ul> <p class=caption><span class=caption-text>User Guide</span></p> <ul class=current> <li class=toctree-l1><a href=../usage/ class="reference internal">Usage</a> </li> <li class=toctree-l1><a href=../output/ class="reference internal">Output</a> </li> <li class="toctree-l1 current"><a class="reference internal current" href=#>Parameters</a> <ul class=current> </ul> </li> <li class=toctree-l1><a href=../genomes/ class="reference internal">Genomes</a> </li> <li class=toctree-l1><a href=../spike-in/ class="reference internal">Spike-in Normalization</a> </li> <li class=toctree-l1><a href=../troubleshooting/ class="reference internal">Troubleshooting</a> </li> </ul> <p class=caption><span class=caption-text>Developer Guide</span></p> <ul> <li class=toctree-l1><a href=../../devs/contributing/ class="reference internal">How to contribute</a> </li> </ul> <ul> <li class=toctree-l1><a href=../../changelog/ class="reference internal">Changelog</a> </li> </ul> </div> </div> </nav> <section data-toggle=wy-nav-shift class=wy-nav-content-wrap> <nav class=wy-nav-top role=navigation aria-label="Mobile navigation menu"> <i data-toggle=wy-nav-top class="fa fa-bars"></i> <a href=../..>CHAMPAGNE</a> </nav> <div class=wy-nav-content> <div class=rst-content><div role=navigation aria-label="breadcrumbs navigation"> <ul class=wy-breadcrumbs> <li><a href=../.. class="icon icon-home" aria-label=Docs></a></li> <li class=breadcrumb-item>User Guide</li> <li class="breadcrumb-item active">Parameters</li> <li class=wy-breadcrumbs-aside> <a href=https://github.com/CCBR/CHAMPAGNE/edit/main/guide/params.md>Edit on CCBR/CHAMPAGNE</a> </li> </ul> <hr> </div> <div role=main class=document itemscope=itemscope itemtype=http://schema.org/Article> <div class=section itemprop=articleBody> <h1 id=how-to-set-pipeline-parameters>How to set pipeline parameters<a class=headerlink href=#how-to-set-pipeline-parameters title="Permanent link">&para;</a></h1> <p>Any parameter can be set via the CLI using two hyphens (<code>--</code>) followed by the parameter name and value. For example:</p> <div class=highlight><pre><span></span><code>champagne<span class=w> </span>run<span class=w> </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/champagne_project<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--input<span class=w> </span>assets/samplesheet_full_mm10.csv<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--contrasts<span class=w> </span>assets/contrasts_full_mm10.csv<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--genome<span class=w> </span>mm10<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--run_gem<span class=w> </span><span class=nb>false</span><span class=w> </span><span class=se>\</span>
<span class=w>    </span>--run_chipseeker<span class=w> </span><span class=nb>false</span><span class=w> </span><span class=se>\</span>
<span class=w>    </span>--run_qc<span class=w> </span><span class=nb>true</span>
</code></pre></div> <p>Alternatively, you can create a YAML file with the parameters you want to set. This is useful for managing multiple parameters or for sharing configurations with others. Here's an example YAML file with some common parameters:</p> <p><code>assets/params.yml</code></p> <div class=highlight><pre><span></span><code><span class=nt>input</span><span class=p>:</span><span class=w> </span><span class=s>&#39;./assets/samplesheet_full_mm10.csv&#39;</span>
<span class=nt>contrasts</span><span class=p>:</span><span class=w> </span><span class=s>&#39;./assets/contrasts_full_mm10.csv&#39;</span>
<span class=nt>genome</span><span class=p>:</span><span class=w> </span><span class="l l-Scalar l-Scalar-Plain">mm10</span>
<span class=nt>run_gem</span><span class=p>:</span><span class=w> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class=nt>run_chipseeker</span><span class=p>:</span><span class=w> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class=nt>run_qc</span><span class=p>:</span><span class=w> </span><span class="l l-Scalar l-Scalar-Plain">true</span>
</code></pre></div> <p>You can then use these parameters with the <code>-params-file</code> option:</p> <div class=highlight><pre><span></span><code>champagne<span class=w> </span>run<span class=w> </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/champagne_project<span class=w> </span><span class=se>\</span>
<span class=w>    </span>-params-file<span class=w> </span>assets/params.yml
</code></pre></div> <p>View the full list of pipeline parameters below.</p> <!-- This doc is generated by: nf-core pipelines schema docs. Edit parameter descriptions in nextflow_schema.json --> <h1 id=ccbrchampagne-pipeline-parameters>CCBR/CHAMPAGNE pipeline parameters<a class=headerlink href=#ccbrchampagne-pipeline-parameters title="Permanent link">&para;</a></h1> <p>CHromAtin iMmuno PrecipitAtion sequencinG aNalysis pipEline</p> <h2 id=inputoutput-options>Input/output options<a class=headerlink href=#inputoutput-options title="Permanent link">&para;</a></h2> <p>The most commonly used pipeline options</p> <table> <thead> <tr> <th>Parameter</th> <th>Description</th> <th>Type</th> <th>Default</th> <th>Required</th> <th>Hidden</th> </tr> </thead> <tbody> <tr> <td><code>input</code></td> <td>Path to comma-separated file containing information about the samples in the experiment. <details><summary>Help</summary><small>You will need to create a design file with information about the samples in your experiment before running the pipeline. Use this parameter to specify its location. It has to be a comma-separated file with 3 columns, and a header row.</small></details></td> <td><code>string</code></td> <td></td> <td>True</td> <td></td> </tr> <tr> <td><code>contrasts</code></td> <td>Optional contrasts specification for differential analysis</td> <td><code>string</code></td> <td></td> <td></td> <td></td> </tr> <tr> <td><code>genome</code></td> <td>Reference genome (e.g. hg38, mm10). This can be a genome in conf/genomes.config, or see 'Custom genome options' to build a custom reference from a fasta &amp; gtf file.</td> <td><code>string</code></td> <td></td> <td>True</td> <td></td> </tr> <tr> <td><code>outputDir</code></td> <td></td> <td><code>string</code></td> <td>${launchDir}/results</td> <td></td> <td>True</td> </tr> <tr> <td><code>tracedir</code></td> <td></td> <td><code>string</code></td> <td>${outputDir}/pipeline_info</td> <td></td> <td>True</td> </tr> <tr> <td><code>publish_dir_mode</code></td> <td></td> <td><code>string</code></td> <td>link</td> <td></td> <td></td> </tr> </tbody> </table> <h2 id=custom-genome-options>Custom genome options<a class=headerlink href=#custom-genome-options title="Permanent link">&para;</a></h2> <p>Use these to build a custom reference genome not already listed in conf/genomes.config. For an example use-case, see conf/test.config.</p> <table> <thead> <tr> <th>Parameter</th> <th>Description</th> <th>Type</th> <th>Default</th> <th>Required</th> <th>Hidden</th> </tr> </thead> <tbody> <tr> <td><code>genome_fasta</code></td> <td>Genome fasta file</td> <td><code>string</code></td> <td></td> <td></td> <td></td> </tr> <tr> <td><code>genes_gtf</code></td> <td>Genome gtf file</td> <td><code>string</code></td> <td></td> <td></td> <td></td> </tr> <tr> <td><code>blacklist</code></td> <td>Blacklisted sequences fasta file</td> <td><code>string</code></td> <td></td> <td></td> <td></td> </tr> <tr> <td><code>rename_contigs</code></td> <td>File with map to translate chromosome names (see assets/R64-1-1_ensembl2UCSC.txt as an example)</td> <td><code>string</code></td> <td></td> <td></td> <td></td> </tr> <tr> <td><code>index_dir</code></td> <td>Absolute path to directory containing pre-built reference genomes</td> <td><code>string</code></td> <td></td> <td></td> <td></td> </tr> </tbody> </table> <h2 id=general-parameters>General parameters<a class=headerlink href=#general-parameters title="Permanent link">&para;</a></h2> <table> <thead> <tr> <th>Parameter</th> <th>Description</th> <th>Type</th> <th>Default</th> <th>Required</th> <th>Hidden</th> </tr> </thead> <tbody> <tr> <td><code>seq_center</code></td> <td></td> <td><code>string</code></td> <td></td> <td></td> <td></td> </tr> <tr> <td><code>read_length</code></td> <td></td> <td><code>integer</code></td> <td></td> <td></td> <td></td> </tr> <tr> <td><code>max_memory</code></td> <td></td> <td><code>string</code></td> <td>224 GB</td> <td></td> <td></td> </tr> <tr> <td><code>max_cpus</code></td> <td></td> <td><code>integer</code></td> <td>32</td> <td></td> <td></td> </tr> <tr> <td><code>max_time</code></td> <td></td> <td><code>string</code></td> <td>72 h</td> <td></td> <td></td> </tr> <tr> <td><code>align_min_quality</code></td> <td></td> <td><code>integer</code></td> <td>6</td> <td></td> <td></td> </tr> <tr> <td><code>min_fragment_length</code></td> <td></td> <td><code>integer</code></td> <td>200</td> <td></td> <td></td> </tr> </tbody> </table> <h2 id=spike-in-options>Spike-in options<a class=headerlink href=#spike-in-options title="Permanent link">&para;</a></h2> <p>Options for experiments that use a spike-in genome</p> <table> <thead> <tr> <th>Parameter</th> <th>Description</th> <th>Type</th> <th>Default</th> <th>Required</th> <th>Hidden</th> </tr> </thead> <tbody> <tr> <td><code>spike_genome</code></td> <td>Optional spike-in genome (e.g. dmelr6.32, ecoli_k12). If null, spike-in normalization will not be performed.</td> <td><code>string</code></td> <td></td> <td></td> <td></td> </tr> <tr> <td><code>spike_norm_method</code></td> <td>Method to compute scaling factors for spike-in normalization. "guenther" uses a simple fraction of the reads aligning to the spike-in genome as described in https://doi.org/10.1016/j.celrep.2014.10.018. "delorenzi" uses deepTools multiBamSummary with --scalingFactors, which is similar to the method described in https://doi.org/10.1101/gr.168260.113.</td> <td><code>string</code></td> <td>guenther</td> <td></td> <td></td> </tr> <tr> <td><code>spike_deeptools_bin_size</code></td> <td>When spike_norm_method is delorenzi, this sets --binSize in deepTools multiBamSummary</td> <td><code>integer</code></td> <td>5000</td> <td></td> <td></td> </tr> <tr> <td><code>spike_deeptools_min_map_quality</code></td> <td>When spike_norm_method is delorenzi, this sets --minMappingQuality in deepTools multiBamSummary</td> <td><code>integer</code></td> <td>30</td> <td></td> <td></td> </tr> </tbody> </table> <h2 id=qc-options>QC options<a class=headerlink href=#qc-options title="Permanent link">&para;</a></h2> <table> <thead> <tr> <th>Parameter</th> <th>Description</th> <th>Type</th> <th>Default</th> <th>Required</th> <th>Hidden</th> </tr> </thead> <tbody> <tr> <td><code>fastq_screen_conf</code></td> <td></td> <td><code>string</code></td> <td></td> <td></td> <td></td> </tr> <tr> <td><code>fastq_screen_db_dir</code></td> <td></td> <td><code>string</code></td> <td></td> <td></td> <td></td> </tr> <tr> <td><code>deeptools_bin_size</code></td> <td></td> <td><code>integer</code></td> <td>25</td> <td></td> <td></td> </tr> <tr> <td><code>deeptools_smooth_length</code></td> <td></td> <td><code>integer</code></td> <td>75</td> <td></td> <td></td> </tr> <tr> <td><code>deeptools_normalize_using</code></td> <td>If using a spike-in genome, recommend setting this to "None"</td> <td><code>string</code></td> <td>RPGC</td> <td></td> <td></td> </tr> <tr> <td><code>deeptools_excluded_chroms</code></td> <td></td> <td><code>string</code></td> <td>chrM chrX chrY</td> <td></td> <td></td> </tr> <tr> <td><code>multiqc_config</code></td> <td></td> <td><code>string</code></td> <td>assets/multiqc_config.yaml</td> <td></td> <td></td> </tr> <tr> <td><code>multiqc_logo</code></td> <td></td> <td><code>string</code></td> <td>assets/ccbr_logo.png</td> <td></td> <td></td> </tr> </tbody> </table> <h2 id=peak-callers>Peak callers<a class=headerlink href=#peak-callers title="Permanent link">&para;</a></h2> <table> <thead> <tr> <th>Parameter</th> <th>Description</th> <th>Type</th> <th>Default</th> <th>Required</th> <th>Hidden</th> </tr> </thead> <tbody> <tr> <td><code>macs_narrow_q</code></td> <td></td> <td><code>number</code></td> <td>0.01</td> <td></td> <td></td> </tr> <tr> <td><code>macs_broad_q</code></td> <td></td> <td><code>number</code></td> <td>0.01</td> <td></td> <td></td> </tr> <tr> <td><code>macs_broad_cutoff</code></td> <td></td> <td><code>number</code></td> <td>0.01</td> <td></td> <td></td> </tr> <tr> <td><code>gem_read_dists</code></td> <td></td> <td><code>string</code></td> <td>assets/gem/Read_Distribution_default.txt</td> <td></td> <td></td> </tr> <tr> <td><code>gem_fold</code></td> <td></td> <td><code>integer</code></td> <td>3</td> <td></td> <td></td> </tr> <tr> <td><code>gem_k_min</code></td> <td></td> <td><code>integer</code></td> <td>6</td> <td></td> <td></td> </tr> <tr> <td><code>gem_k_max</code></td> <td></td> <td><code>integer</code></td> <td>13</td> <td></td> <td></td> </tr> <tr> <td><code>sicer_species</code></td> <td></td> <td><code>string</code></td> <td></td> <td></td> <td></td> </tr> </tbody> </table> <h2 id=motifs>motifs<a class=headerlink href=#motifs title="Permanent link">&para;</a></h2> <table> <thead> <tr> <th>Parameter</th> <th>Description</th> <th>Type</th> <th>Default</th> <th>Required</th> <th>Hidden</th> </tr> </thead> <tbody> <tr> <td><code>homer_de_novo</code></td> <td></td> <td><code>boolean</code></td> <td>True</td> <td></td> <td></td> </tr> <tr> <td><code>homer_jaspar_db</code></td> <td></td> <td><code>string</code></td> <td>assets/JASPAR2022_CORE_vertebrates_non-redundant_pfms_jaspar.txt</td> <td></td> <td></td> </tr> </tbody> </table> <h2 id=run-control>run control<a class=headerlink href=#run-control title="Permanent link">&para;</a></h2> <p>Toggle various steps of the pipeline on/off</p> <table> <thead> <tr> <th>Parameter</th> <th>Description</th> <th>Type</th> <th>Default</th> <th>Required</th> <th>Hidden</th> </tr> </thead> <tbody> <tr> <td><code>run_qc</code></td> <td></td> <td><code>boolean</code></td> <td>True</td> <td></td> <td></td> </tr> <tr> <td><code>run_deeptools</code></td> <td></td> <td><code>boolean</code></td> <td>True</td> <td></td> <td></td> </tr> <tr> <td><code>run_normalize_input</code></td> <td></td> <td><code>boolean</code></td> <td>True</td> <td></td> <td></td> </tr> <tr> <td><code>run_call_peaks</code></td> <td></td> <td><code>boolean</code></td> <td>True</td> <td></td> <td></td> </tr> <tr> <td><code>run_gem</code></td> <td></td> <td><code>boolean</code></td> <td>True</td> <td></td> <td></td> </tr> <tr> <td><code>run_sicer</code></td> <td></td> <td><code>boolean</code></td> <td>True</td> <td></td> <td></td> </tr> <tr> <td><code>run_macs_broad</code></td> <td></td> <td><code>boolean</code></td> <td>True</td> <td></td> <td></td> </tr> <tr> <td><code>run_macs_narrow</code></td> <td></td> <td><code>boolean</code></td> <td>True</td> <td></td> <td></td> </tr> <tr> <td><code>run_normalize_peaks</code></td> <td></td> <td><code>boolean</code></td> <td></td> <td></td> <td></td> </tr> <tr> <td><code>run_chipseeker</code></td> <td></td> <td><code>boolean</code></td> <td></td> <td></td> <td></td> </tr> <tr> <td><code>run_homer</code></td> <td></td> <td><code>boolean</code></td> <td>True</td> <td></td> <td></td> </tr> <tr> <td><code>run_meme</code></td> <td></td> <td><code>boolean</code></td> <td>True</td> <td></td> <td></td> </tr> <tr> <td><code>run_consensus_union</code></td> <td></td> <td><code>boolean</code></td> <td>True</td> <td></td> <td></td> </tr> <tr> <td><code>run_consensus_corces</code></td> <td></td> <td><code>boolean</code></td> <td>True</td> <td></td> <td></td> </tr> </tbody> </table> <h2 id=containers>containers<a class=headerlink href=#containers title="Permanent link">&para;</a></h2> <table> <thead> <tr> <th>Parameter</th> <th>Description</th> <th>Type</th> <th>Default</th> <th>Required</th> <th>Hidden</th> </tr> </thead> <tbody> <tr> <td><code>containers_base</code></td> <td></td> <td><code>string</code></td> <td>nciccbr/ccbr_ubuntu_base_20.04:v6.1</td> <td></td> <td></td> </tr> <tr> <td><code>containers_deeptools</code></td> <td></td> <td><code>string</code></td> <td>nciccbr/ccbr_deeptools_3.5.3:v1</td> <td></td> <td></td> </tr> <tr> <td><code>containers_fastqc</code></td> <td></td> <td><code>string</code></td> <td>nciccbr/ccrgb_qctools:v4.0</td> <td></td> <td></td> </tr> <tr> <td><code>containers_fastq_screen</code></td> <td></td> <td><code>string</code></td> <td>nciccbr/ccbr_fastq_screen_0.14.1:v1.0</td> <td></td> <td></td> </tr> <tr> <td><code>containers_frip</code></td> <td></td> <td><code>string</code></td> <td>nciccbr/ccbr_frip:v1</td> <td></td> <td></td> </tr> <tr> <td><code>containers_gem</code></td> <td></td> <td><code>string</code></td> <td>nciccbr/ccbr_gem_3.4:v1</td> <td></td> <td></td> </tr> <tr> <td><code>containers_macs2</code></td> <td></td> <td><code>string</code></td> <td>nciccbr/ccbr_macs2_2.2.9.1:v1</td> <td></td> <td></td> </tr> <tr> <td><code>containers_multiqc</code></td> <td></td> <td><code>string</code></td> <td>nciccbr/ccbr_multiqc_1.15:v1</td> <td></td> <td></td> </tr> <tr> <td><code>containers_ngsqc</code></td> <td></td> <td><code>string</code></td> <td>nciccbr/ccbr_ngsqc_0.31:v1</td> <td></td> <td></td> </tr> <tr> <td><code>containers_phantom_peaks</code></td> <td></td> <td><code>string</code></td> <td>quay.io/biocontainers/phantompeakqualtools:1.2.2--hdfd78af_1</td> <td></td> <td></td> </tr> <tr> <td><code>containers_picard</code></td> <td></td> <td><code>string</code></td> <td>nciccbr/ccbr_picard_2.27.5:v1</td> <td></td> <td></td> </tr> <tr> <td><code>containers_preseq</code></td> <td></td> <td><code>string</code></td> <td>nciccbr/ccbr_preseq_v2.0:v1</td> <td></td> <td></td> </tr> <tr> <td><code>containers_r</code></td> <td></td> <td><code>string</code></td> <td>nciccbr/ccbr_r_4.3.0:v1</td> <td></td> <td></td> </tr> <tr> <td><code>containers_sicer</code></td> <td></td> <td><code>string</code></td> <td>nciccbr/ccbr_sicer2_1.0.3:v1</td> <td></td> <td></td> </tr> </tbody> </table> <h2 id=other-parameters>Other parameters<a class=headerlink href=#other-parameters title="Permanent link">&para;</a></h2> <table> <thead> <tr> <th>Parameter</th> <th>Description</th> <th>Type</th> <th>Default</th> <th>Required</th> <th>Hidden</th> </tr> </thead> <tbody> <tr> <td><code>diffbind_report</code></td> <td></td> <td><code>string</code></td> <td>assets/diffbind_report.Rmd</td> <td></td> <td>True</td> </tr> </tbody> </table> </div> </div><script async src="https://www.googletagmanager.com/gtag/js?id=G-D3SL9V30KL"></script> <script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-D3SL9V30KL'); // Replace with your tracking ID
</script> </div> </div> </section> </div> <div class=rst-versions role=note aria-label=Versions> <span class=rst-current-version data-toggle=rst-current-version> <span> <a href=https://github.com/CCBR/CHAMPAGNE class="fa fa-code-fork" style="color: #fcfcfc"> CCBR/CHAMPAGNE</a> </span> <span><a href=../output/ style="color: #fcfcfc">&laquo; Previous</a></span> <span><a href=../genomes/ style="color: #fcfcfc">Next &raquo;</a></span> </span> </div> <script src=../../js/jquery-3.6.0.min.js></script> <script>var base_url = "../..";</script> <script src=../../js/theme_extra.js></script> <script src=../../js/theme.js></script> <script src=../../js/custom.js></script> <script src=../../search/main.js></script> <script src=../../js/version-select.js></script> <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script> </body> </html> 