<!DOCTYPE html><html class=writer-html5 lang=en> <head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width, initial-scale=1.0"><meta name=author content="CCR Collaborative Bioinformatics Resource"><link rel="shortcut icon" href=../img/favicon.ico><title>Changelog - CHAMPAGNE</title><link rel=stylesheet href=../css/theme.css><link rel=stylesheet href=../css/theme_extra.css><link rel=stylesheet href=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css><link href=../css/custom.css rel=stylesheet><link href=../css/version-select.css rel=stylesheet><script>
        // Current page data
        var mkdocs_page_name = "Changelog";
        var mkdocs_page_input_path = "changelog.md";
        var mkdocs_page_url = null;
      </script><!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]--><script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js></script><script>hljs.highlightAll();</script></head> <body class=wy-body-for-nav role=document> <div class=wy-grid-for-nav> <nav data-toggle=wy-nav-shift class="wy-nav-side stickynav"> <div class=wy-side-scroll> <div class=wy-side-nav-search> <a href=.. class="icon icon-home"> CHAMPAGNE </a><div role=search> <form id=rtd-search-form class=wy-form action=../search.html method=get> <input type=text name=q placeholder="Search docs" aria-label="Search docs" title="Type search term here"> </form> </div> </div> <div class="wy-menu wy-menu-vertical" data-spy=affix role=navigation aria-label="Navigation menu"> <ul> <li class=toctree-l1><a class="reference internal" href=..>Home</a> </li> </ul> <p class=caption><span class=caption-text>User Guide</span></p> <ul> <li class=toctree-l1><a href=../guide/usage/ class="reference internal">Usage</a> </li> <li class=toctree-l1><a href=../guide/output/ class="reference internal">Output</a> </li> <li class=toctree-l1><a href=../guide/params/ class="reference internal">Parameters</a> </li> <li class=toctree-l1><a href=../guide/genomes/ class="reference internal">Genomes</a> </li> <li class=toctree-l1><a href=../guide/spike-in/ class="reference internal">Spike-in Normalization</a> </li> <li class=toctree-l1><a href=../guide/troubleshooting/ class="reference internal">Troubleshooting</a> </li> </ul> <p class=caption><span class=caption-text>Developer Guide</span></p> <ul> <li class=toctree-l1><a href=../devs/contributing/ class="reference internal">How to contribute</a> </li> </ul> <ul class=current> <li class="toctree-l1 current"><a class="reference internal current" href=#>Changelog</a> <ul class=current> <li class=toctree-l2><a class="reference internal" href=#new-features>New features</a> </li> <li class=toctree-l2><a class="reference internal" href=#bug-fixes>Bug fixes</a> </li> <li class=toctree-l2><a class="reference internal" href=#documentation>Documentation</a> </li> </ul> </li> </ul> </div> </div> </nav> <section data-toggle=wy-nav-shift class=wy-nav-content-wrap> <nav class=wy-nav-top role=navigation aria-label="Mobile navigation menu"> <i data-toggle=wy-nav-top class="fa fa-bars"></i> <a href=..>CHAMPAGNE</a> </nav> <div class=wy-nav-content> <div class=rst-content><div role=navigation aria-label="breadcrumbs navigation"> <ul class=wy-breadcrumbs> <li><a href=.. class="icon icon-home" aria-label=Docs></a></li> <li class="breadcrumb-item active">Changelog</li> <li class=wy-breadcrumbs-aside> <a href=https://github.com/CCBR/CHAMPAGNE/edit/main/changelog.md>Edit on CCBR/CHAMPAGNE</a> </li> </ul> <hr> </div> <div role=main class=document itemscope=itemscope itemtype=http://schema.org/Article> <div class=section itemprop=articleBody> <h2 id=champagne-development-version>CHAMPAGNE development version<a class=headerlink href=#champagne-development-version title="Permanent link">&para;</a></h2> <ul> <li>CHAMPAGNE now depends on ccbr_tools v0.4 for updated jobby &amp; spooker utilities. (#247, @kelly-sovacool)</li> <li>Reorganized output directory structure. (#256, @kelly-sovacool)</li> <li>Nextflow parameters are now validated at the start of the workflow. (#257, @kelly-sovacool)</li> </ul> <h3 id=new-features>New features<a class=headerlink href=#new-features title="Permanent link">&para;</a></h3> <ul> <li>Contrasts are now specified as a TSV file instead of YAML. (#224, @kelly-sovacool)</li> <li>See the example contrast sheets in <code>assets/</code>.</li> <li>New consensus peak method from Corces <em>et al.</em> (<a href=https://www.science.org/doi/10.1126/science.aav1898>doi:10.1126/science.aav1898</a>). (#225, #246, @kelly-sovacool)</li> <li>Normalization with a spike-in genome is now supported. (#259, @kelly-sovacool)</li> <li>Set <code>params.spike_genome</code> to <code>dmelr6.32</code> or <code>ecoli_k12</code></li> <li>CLI improvements</li> <li>Use <code>nextflow run -resume</code> by default, or turn it off with <code>champagne run --forceall</code>. (#224, @kelly-sovacool)</li> <li>Add <code>--output</code> argument for <code>champagne init</code> and <code>champagne run</code>. (#232, #233, @kelly-sovacool)<ul> <li>This is equivalent to the nextflow launchDir constant.</li> </ul> </li> <li><code>--mode</code> now has <code>slurm</code> as the default. (#249, @kelly-sovacool)</li> <li>Improved nextflow options:</li> <li>Set <code>publish_dir_mode</code> to <code>link</code> by default.</li> <li>Set <code>process.cache</code> to <code>deep</code> by default rather than lenient. (#224, @kelly-sovacool)</li> <li>Enable the nextflow timeline &amp; trace reports by default. (#226, @kelly-sovacool)</li> <li>Output BioCompute objects by default using the nf-prov plugin. (#260, @kelly-sovacool)</li> </ul> <h3 id=bug-fixes>Bug fixes<a class=headerlink href=#bug-fixes title="Permanent link">&para;</a></h3> <ul> <li>Refactor checks for the sample sheet &amp; contrast sheet to prevent unnecessary re-running. (#224, @kelly-sovacool)</li> <li>Fix a file name clash during input pooling. (#224, @kelly-sovacool)</li> <li>Fix bug in MEME AME process that caused it not to run on all samples. (#234, @kelly-sovacool)</li> <li>Also correct the motif rank calculation. (#234, @kopardev)</li> <li>fix: ensure inputs are pooled correctly for paired end reads. (#250, @kelly-sovacool)</li> <li>fix chipseeker filename collision. (#252, @kelly-sovacool)</li> </ul> <h3 id=documentation>Documentation<a class=headerlink href=#documentation title="Permanent link">&para;</a></h3> <ul> <li>Now using the readthedocs theme for the docs website. (#236, @kelly-sovacool)</li> <li>Improved help message for <code>champagne run</code>. (#249, @kelly-sovacool)</li> <li>Major improvements to fill out the documentation website. (#262, @kelly-sovacool)</li> </ul> <h2 id=champagne-041>CHAMPAGNE 0.4.1<a class=headerlink href=#champagne-041 title="Permanent link">&para;</a></h2> <ul> <li>The CHAMPAGNE nextflow workflow now has a version entry in <code>nextflow.config</code>, in compliance with nf-core. (#213, @kelly-sovacool)</li> <li>Pool input (control) reads of the same sample name by default. Any inputs that should not be pooled must have different sample names in the samplesheet. (#214, @kelly-sovacool)</li> <li>Add histone samples to the <code>test_human</code> dataset. (#215, @kelly-sovacool)</li> </ul> <h2 id=champagne-040>CHAMPAGNE 0.4.0<a class=headerlink href=#champagne-040 title="Permanent link">&para;</a></h2> <h3 id=new-features_1>New features<a class=headerlink href=#new-features_1 title="Permanent link">&para;</a></h3> <ul> <li>Create a script (<code>bin/champagne</code>) to provide an interface to the champagne CLI that works out-of-the-box without the need to install the python package with <code>pip</code>. (#180, @kelly-sovacool)</li> <li>However, any dependencies not in the Python Standard Library must be installed for this to work. See the dependencies list in <code>pyproject.toml</code>.</li> <li>Allow additional columns in the sample sheet beyond the minimum required header. (#176, @kelly-sovacool)</li> <li>Add a workflow entry point to download fastq files from SRA. (#176, @kelly-sovacool)</li> <li>Add <code>test_human</code> profile with chipseq data from ENCODE. (#176, @kelly-sovacool)</li> </ul> <h3 id=bug-fixes_1>Bug fixes<a class=headerlink href=#bug-fixes_1 title="Permanent link">&para;</a></h3> <ul> <li>Fix configuration files for compatibility with using the GitHub repo as the source. (#173, @kelly-sovacool)</li> <li>These equivalent commands now work: <div class=highlight><pre><span></span><code>nextflow<span class=w> </span>run<span class=w> </span>CCBR/CHAMPAGNE
champagne<span class=w> </span>run<span class=w> </span>--main<span class=w> </span>CCBR/CHAMPAGNE
</code></pre></div></li> <li>Allow multiple samples to use the same input. (#176, @kelly-sovacool)</li> <li>In the biowulf config profile, switch variable $SLURM_JOBID to $SLURM_JOB_ID. (@kelly-sovacool)</li> <li>Increase resource allocations for chipseeker and deeptools. (#192, @slsevilla)</li> <li>Check the validity of the contrastsheet earlier on in the workflow. (#192, @slsevilla; #200, @kelly-sovacool)</li> <li>Fix bug where <code>manorm</code> was using R1 twice instead of R1 and R2. (#206, @kelly-sovacool)</li> </ul> <h3 id=misc>Misc<a class=headerlink href=#misc title="Permanent link">&para;</a></h3> <ul> <li>Change the peak widths histogram type from overlay to stack. (#176, @kelly-sovacool)</li> <li>Documentation improvements. (#192, @slsevilla)</li> </ul> <h2 id=champagne-030>CHAMPAGNE 0.3.0<a class=headerlink href=#champagne-030 title="Permanent link">&para;</a></h2> <h3 id=new-features_2>New features<a class=headerlink href=#new-features_2 title="Permanent link">&para;</a></h3> <ul> <li>Find motifs in the genome with Homer. (#142)</li> <li>Run motif enrichment analysis with MEME. (#142)</li> <li>Annotate peaks with chipseeker. (#142,#147,#157)</li> <li>Add preseq complexity curve and fastq screen to multiqc report. (#147)</li> <li>Support multiple replicates per sample and call consensus peaks on replicates. (#129)</li> <li>Optionally normalize p-values with the <a href=https://github.com/CCBR/nf-modules/tree/60d50f4c45a50378cad70b49013f51750617caaa/subworkflows/CCBR/consensus_peaks>CCBR/consensus_peaks</a> subworkflow.</li> <li>Implement differential peak calling. (#158)</li> <li>Optionally specify contrasts via a YAML file. If no file is specified, differential analysis is not performed.</li> <li>If any sample has only one replicate, run <code>MAnorm</code>, otherwise run <code>diffbind</code>.</li> <li>Print the recommended citation in bibtex format with <code>champagne --citation</code>. (#153)</li> <li>CHAMPAGNE is also now archived in Zenodo with DOI <code>10.5281/zenodo.10516078</code>.</li> <li>The docs website now has a dropdown menu to select which version to view. The latest release is shown by default. (#170)</li> </ul> <h3 id=bug-fixes_2>Bug fixes<a class=headerlink href=#bug-fixes_2 title="Permanent link">&para;</a></h3> <ul> <li>Fix deepTools plots (#144):</li> <li>Per sample fingerprint plots instead of per replicate.</li> <li>Input normalized profile plots.</li> <li>Protein-coding-only versions of plots.</li> <li>Ensure sample IDs are sorted. (#150)</li> <li>Fix a bug where the wrong SICER output file was used for downstream analyses. (#155)</li> <li>Fix CLI profile on machines other than biowulf &amp; FRCE. (#168)</li> <li>Fix broken bold styling in documentation website. (#53)</li> </ul> <h2 id=champagne-022>CHAMPAGNE 0.2.2<a class=headerlink href=#champagne-022 title="Permanent link">&para;</a></h2> <ul> <li>Fix permissions issues in the CLI. (#167)</li> </ul> <h2 id=champagne-021>CHAMPAGNE 0.2.1<a class=headerlink href=#champagne-021 title="Permanent link">&para;</a></h2> <ul> <li>Fix a bug in QC stats that mixed up the statistics for different samples. (#125)</li> <li>Fix a bug in the CLI that added the <code>-profile</code> to the nextflow command even if it wasn't needed (#125).</li> <li>Report read counts between blacklist &amp; filtering steps in the QC table. (#125)</li> <li>Run spooker on workflow completion (#126).</li> </ul> <h2 id=champagne-020>CHAMPAGNE 0.2.0<a class=headerlink href=#champagne-020 title="Permanent link">&para;</a></h2> <h3 id=new-features_3>New features<a class=headerlink href=#new-features_3 title="Permanent link">&para;</a></h3> <ul> <li>Implement peak calling with sicer2, macs2, and gem. (#52)</li> <li>Add parameter options to skip QC, input normalization, and/or peak calling steps. (#72)</li> <li>Calculate and plot QC metrics for called peaks:</li> <li>Fraction in Peaks (FRiP) (#89)</li> <li>Jaccard index (#92)</li> <li>Histogram of peak widths (#92)</li> <li>Add support for paired-end reads. (#105)</li> <li>Add an option to use a custom reference from a genome fasta, gtf, and blacklist file. (#105)</li> <li>Champagne CLI: (#112)</li> <li>New <code>--mode</code> option for <code>champagne run</code> to execute the workflow locally ('local') or submit it as a slurm job ('slurm').</li> <li>Option to override the path to the champagne <code>main.nf</code> file or specify the github repo (<code>CCBR/CHAMPAGNE</code>) instead. <div class=highlight><pre><span></span><code><span class=c1># use the default path</span>
champagne<span class=w> </span>run<span class=w> </span>...
<span class=c1># override the path</span>
champagne<span class=w> </span>run<span class=w> </span>path/to/champagne/main.nf
<span class=c1># use a revision from github instead</span>
champagne<span class=w> </span>run<span class=w> </span>CCBR/CHAMPAGNE<span class=w> </span>-r<span class=w> </span>v0.1.0
</code></pre></div></li> </ul> <h3 id=bug-fixes_3>Bug fixes<a class=headerlink href=#bug-fixes_3 title="Permanent link">&para;</a></h3> <ul> <li>CLI:</li> <li>Error when biowulf-specific environment variables are not defined. (#54)</li> <li>The host is now correctly detected as biowulf via <code>scontrol</code>. (#75)</li> <li>Containers:</li> <li>Containers are now specified in process definitions instead of <code>withName</code>/<code>withLabel</code> for better control. (#69)<ul> <li>Shared containers are specified as parameters in the config file <code>conf/containers.config</code>.</li> </ul> </li> <li>No longer use <code>--mount type=bind</code> or <code>--volume</code> for making directories available to processes in containers. Instead, use Nextflow's <code>Channel.fromPath</code> constructor with <code>type: 'dir'</code>. (#71)</li> </ul> <h3 id=api-breaking-changes>API-breaking changes<a class=headerlink href=#api-breaking-changes title="Permanent link">&para;</a></h3> <ul> <li>An error is thrown when a required input file doesn't exist. (#71)</li> <li>Previously, the workflow quietly didn't run the process(es) that required the missing file.</li> <li>Renamed <code>champagne config</code> to <code>champagne init</code> to avoid clashing with <code>nextflow config</code>. (#112)</li> </ul> <h2 id=champagne-010>CHAMPAGNE 0.1.0<a class=headerlink href=#champagne-010 title="Permanent link">&para;</a></h2> <h3 id=quality-control-steps-implemented-for-single-end-reads>Quality control steps implemented for single-end reads<a class=headerlink href=#quality-control-steps-implemented-for-single-end-reads title="Permanent link">&para;</a></h3> <ul> <li>Trim raw reads, FastQC on raw and trimmed reads, and FastQ Screen on trimmed reads.</li> <li>Exclude reads that align to blacklist regions, align remaining reads to the reference genome, and deduplicate.</li> <li>Preseq on aligned reads.</li> <li>Phantompeakqualtools on aligned and deduplicated reads.</li> <li>Process reads with deepTools: bam coverage to generate bigwigs for each sample, summarize all bigwigs, and compute matrices relative to TSSs and scaled to metagene regions.</li> <li>Generate plots with deepTools: PCA, profile, heatmap, spearman correlation, and fingerprint plots.</li> <li>Summarize all quality control steps in a MultiQC report.</li> <li>Input-normalize ChIP fragments for the next stage of the pipeline.</li> </ul> <!-- the contents are automatically inserted here. do not edit this file, instead edit the file specified above to make changes. --> </div> </div><script async src="https://www.googletagmanager.com/gtag/js?id=G-D3SL9V30KL"></script> <script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-D3SL9V30KL'); // Replace with your tracking ID
</script> </div> </div> </section> </div> <div class=rst-versions role=note aria-label=Versions> <span class=rst-current-version data-toggle=rst-current-version> <span> <a href=https://github.com/CCBR/CHAMPAGNE class="fa fa-code-fork" style="color: #fcfcfc"> CCBR/CHAMPAGNE</a> </span> <span><a href=../devs/contributing/ style="color: #fcfcfc">&laquo; Previous</a></span> </span> </div> <script src=../js/jquery-3.6.0.min.js></script> <script>var base_url = "..";</script> <script src=../js/theme_extra.js></script> <script src=../js/theme.js></script> <script src=../js/custom.js></script> <script src=../search/main.js></script> <script src=../js/version-select.js></script> <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script> </body> </html> 