<!DOCTYPE html><html class=writer-html5 lang=en> <head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width, initial-scale=1.0"><meta name=author content="CCR Collaborative Bioinformatics Resource"><link rel="shortcut icon" href=../../img/favicon.ico><title>Genomes - CHAMPAGNE</title><link rel=stylesheet href=../../css/theme.css><link rel=stylesheet href=../../css/theme_extra.css><link rel=stylesheet href=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css><link href=../../css/custom.css rel=stylesheet><link href=../../css/version-select.css rel=stylesheet><script>
        // Current page data
        var mkdocs_page_name = "Genomes";
        var mkdocs_page_input_path = "guide/genomes.md";
        var mkdocs_page_url = null;
      </script><!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]--><script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js></script><script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/yaml.min.js></script><script>hljs.highlightAll();</script></head> <body class=wy-body-for-nav role=document> <div class=wy-grid-for-nav> <nav data-toggle=wy-nav-shift class="wy-nav-side stickynav"> <div class=wy-side-scroll> <div class=wy-side-nav-search> <a href=../.. class="icon icon-home"> CHAMPAGNE </a><div role=search> <form id=rtd-search-form class=wy-form action=../../search.html method=get> <input type=text name=q placeholder="Search docs" aria-label="Search docs" title="Type search term here"> </form> </div> </div> <div class="wy-menu wy-menu-vertical" data-spy=affix role=navigation aria-label="Navigation menu"> <ul> <li class=toctree-l1><a class="reference internal" href=../..>Home</a> </li> </ul> <p class=caption><span class=caption-text>User Guide</span></p> <ul class=current> <li class=toctree-l1><a href=../usage/ class="reference internal">Usage</a> </li> <li class=toctree-l1><a href=../output/ class="reference internal">Output</a> </li> <li class=toctree-l1><a href=../params/ class="reference internal">Parameters</a> </li> <li class="toctree-l1 current"><a class="reference internal current" href=#>Genomes</a> <ul class=current> <li class=toctree-l2><a class="reference internal" href=#supported-reference-genomes>Supported reference genomes</a> <ul> <li class=toctree-l3><a class="reference internal" href=#reference-genomes>Reference Genomes</a> <ul> <li class=toctree-l4><a class="reference internal" href=#hg38>hg38</a> </li> <li class=toctree-l4><a class="reference internal" href=#hg19>hg19</a> </li> <li class=toctree-l4><a class="reference internal" href=#mm10>mm10</a> </li> </ul> </li> <li class=toctree-l3><a class="reference internal" href=#spike-in-genomes>Spike-in Genomes</a> <ul> <li class=toctree-l4><a class="reference internal" href=#dmelr632>dmelr6.32</a> </li> <li class=toctree-l4><a class="reference internal" href=#ecoli_k12>ecoli_k12</a> </li> </ul> </li> <li class=toctree-l3><a class="reference internal" href=#custom-blacklist>Custom blacklist</a> </li> <li class=toctree-l3><a class="reference internal" href=#custom-reference-genome>Custom reference genome</a> </li> </ul> </li> </ul> </li> <li class=toctree-l1><a href=../spike-in/ class="reference internal">Spike-in Normalization</a> </li> <li class=toctree-l1><a href=../troubleshooting/ class="reference internal">Troubleshooting</a> </li> </ul> <p class=caption><span class=caption-text>Developer Guide</span></p> <ul> <li class=toctree-l1><a href=../../devs/contributing/ class="reference internal">How to contribute</a> </li> <li class=toctree-l1><a href=../../devs/contributors/ class="reference internal">Contributors</a> </li> </ul> <ul> <li class=toctree-l1><a href=../../changelog/ class="reference internal">Changelog</a> </li> </ul> </div> </div> </nav> <section data-toggle=wy-nav-shift class=wy-nav-content-wrap> <nav class=wy-nav-top role=navigation aria-label="Mobile navigation menu"> <i data-toggle=wy-nav-top class="fa fa-bars"></i> <a href=../..>CHAMPAGNE</a> </nav> <div class=wy-nav-content> <div class=rst-content><div role=navigation aria-label="breadcrumbs navigation"> <ul class=wy-breadcrumbs> <li><a href=../.. class="icon icon-home" aria-label=Docs></a></li> <li class=breadcrumb-item>User Guide</li> <li class="breadcrumb-item active">Genomes</li> <li class=wy-breadcrumbs-aside> <a href=https://github.com/CCBR/CHAMPAGNE/edit/main/guide/genomes.md>Edit on CCBR/CHAMPAGNE</a> </li> </ul> <hr> </div> <div role=main class=document itemscope=itemscope itemtype=http://schema.org/Article> <div class=section itemprop=articleBody> <h1 id=genomes>Genomes<a class=headerlink href=#genomes title="Permanent link">&para;</a></h1> <!--
This file is created by concatenating _genomes_tail.md and the auto-generated genomes list.
Do not edit guide/genomes.md manually.
--> <h2 id=supported-reference-genomes>Supported reference genomes<a class=headerlink href=#supported-reference-genomes title="Permanent link">&para;</a></h2> <p>These genomes are available on biowulf.</p> <h3 id=reference-genomes>Reference Genomes<a class=headerlink href=#reference-genomes title="Permanent link">&para;</a></h3> <p>These genomes can be passed to the <code>--genome</code> parameter.</p> <h4 id=hg38><code>hg38</code><a class=headerlink href=#hg38 title="Permanent link">&para;</a></h4> <ul> <li>species: <code>Homo sapiens</code></li> <li>fasta: <code>${params.index_dir}/hg38_basic/hg38_basic.fa</code></li> <li>genes_gtf: <code>${params.index_dir}/hg38_basic/genes.gtf</code></li> <li>blacklist_index: <code>${params.index_dir}/hg38_basic/indexes/blacklist/hg38.blacklist_v3.chrM.chr_rDNA.*</code></li> <li>reference_index: <code>${params.index_dir}/hg38_basic/bwa_index/hg38*</code></li> <li>chromosomes_dir: <code>${params.index_dir}/hg38_basic/Chromsomes/</code></li> <li>chrom_sizes: <code>${params.index_dir}/hg38_basic/indexes/hg38.fa.sizes</code></li> <li>gene_info: <code>${params.index_dir}/hg38_basic/geneinfo.bed</code></li> <li>effective_genome_size: <code>2700000000</code></li> <li>meme_motifs: <code>${projectDir}/assets/HOCOMOCOv11_core_HUMAN_mono_meme_format.tar.gz</code></li> <li>bioc_txdb: <code>TxDb.Hsapiens.UCSC.hg38.knownGene</code></li> <li>bioc_annot: <code>org.Hs.eg.db</code></li> </ul> <h4 id=hg19><code>hg19</code><a class=headerlink href=#hg19 title="Permanent link">&para;</a></h4> <ul> <li>species: <code>Homo sapiens</code></li> <li>fasta: <code>${params.index_dir}/hg19_basic/hg19.fa</code></li> <li>genes_gtf: <code>${params.index_dir}/hg19_basic/genes.gtf</code></li> <li>blacklist_index: <code>${params.index_dir}/hg19_basic/indexes/blacklist/hg19.blacklist.*</code></li> <li>reference_index: <code>${params.index_dir}/hg19_basic/bwa_index/hg19.*</code></li> <li>chromosomes_dir: <code>${params.index_dir}/hg19_basic/Chromosomes</code></li> <li>chrom_sizes: <code>${params.index_dir}/hg19_basic/Chromosomes/chrom.sizes</code></li> <li>gene_info: <code>${params.index_dir}/hg19_basic/geneinfo.bed</code></li> <li>effective_genome_size: <code>2700000000</code></li> <li>meme_motifs: <code>${projectDir}/assets/HOCOMOCOv11_core_HUMAN_mono_meme_format.tar.gz</code></li> <li>bioc_txdb: <code>TxDb.Hsapiens.UCSC.hg38.knownGene</code></li> <li>bioc_annot: <code>org.Hs.eg.db</code></li> </ul> <h4 id=mm10><code>mm10</code><a class=headerlink href=#mm10 title="Permanent link">&para;</a></h4> <ul> <li>species: <code>Mus musculus</code></li> <li>fasta: <code>${params.index_dir}/mm10_basic/mm10_basic.fa</code></li> <li>genes_gtf: <code>${params.index_dir}/mm10_basic/genes.gtf</code></li> <li>blacklist_index: <code>${params.index_dir}/mm10_basic/indexes/blacklist/mm10.blacklist.chrM.chr_rDNA.*</code></li> <li>reference_index: <code>${params.index_dir}/mm10_basic/indexes/reference/mm10*</code></li> <li>chromosomes_dir: <code>${params.index_dir}/mm10_basic/Chromsomes/</code></li> <li>chrom_sizes: <code>${params.index_dir}/mm10_basic/indexes/mm10.fa.sizes</code></li> <li>gene_info: <code>${params.index_dir}/mm10_basic/geneinfo.bed</code></li> <li>effective_genome_size: <code>2400000000</code></li> <li>meme_motifs: <code>${projectDir}/assets/HOCOMOCOv11_core_MOUSE_mono_meme_format.tar.gz</code></li> <li>bioc_txdb: <code>TxDb.Mmusculus.UCSC.mm10.knownGene</code></li> <li>bioc_annot: <code>org.Mmu.eg.db</code></li> </ul> <h3 id=spike-in-genomes>Spike-in Genomes<a class=headerlink href=#spike-in-genomes title="Permanent link">&para;</a></h3> <p>These genomes can be passed to the <code>--spike_genome</code> parameter.</p> <h4 id=dmelr632><code>dmelr6.32</code><a class=headerlink href=#dmelr632 title="Permanent link">&para;</a></h4> <ul> <li>species: <code>Drosophila melanogaster</code></li> <li>fasta: <code>${params.index_dir}/dmelr6.32/indexes/dmelr6.32.fa</code></li> <li>reference_index: <code>${params.index_dir}/dmelr6.32/indexes/bwa/Drosophila_melanogaster.*</code></li> <li>blacklist_index: <code>${params.index_dir}/dmelr6.32/indexes/blacklist/dmelr6.32.blacklist.*</code></li> <li>blacklist_bed: <code>${params.index_dir}/dmelr6.32/indexes/dm6-blacklist.v2.no_chr.bed.gz</code></li> <li>effective_genome_size: <code>142573017</code></li> <li>chrom_sizes: <code>${params.index_dir}/dmelr6.32/chrom_dmelr6.28.sizes</code></li> </ul> <h4 id=ecoli_k12><code>ecoli_k12</code><a class=headerlink href=#ecoli_k12 title="Permanent link">&para;</a></h4> <ul> <li>species: <code>Escherichia coli K-12</code></li> <li>fasta: <code>${params.index_dir}/ecoli_k12/indexes/ecoli_k12.fa</code></li> <li>reference_index: <code>${params.index_dir}/ecoli_k12/indexes/ecoli_k12.*</code></li> <li>blacklist_bed: <code>NO_FILE</code></li> <li>effective_genome_size: <code>4641652</code></li> <li>chrom_sizes: <code>${params.index_dir}/ecoli_k12/Chromosomes/chrom.sizes</code></li> </ul> <h3 id=custom-blacklist>Custom blacklist<a class=headerlink href=#custom-blacklist title="Permanent link">&para;</a></h3> <p>If you'd like to override the default blacklist used by one of the built-in genomes, you can provide a custom blacklist bed file or fasta file:</p> <div class=highlight><pre><span></span><code>champagne<span class=w> </span>run<span class=w> </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/champagne_project<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--mode<span class=w> </span>slurm<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--genome<span class=w> </span>hg38<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--blacklist<span class=w> </span>/path/to/blacklist.bed
</code></pre></div> <p>If you're providing a custom blacklist bed file, make sure its regions refer to the genome version you're using.</p> <h3 id=custom-reference-genome>Custom reference genome<a class=headerlink href=#custom-reference-genome title="Permanent link">&para;</a></h3> <p>If you'd like to use a genome not available on Biowulf, you can prepare a custom genome with the <code>MAKE_REFERENCE</code> entrypoint. If you'd like to use a custom genome, you'll need the following files:</p> <ul> <li>genome fasta</li> <li>genome GTF</li> <li>blacklist fasta</li> </ul> <p>Prepare your custom reference genome with:</p> <div class=highlight><pre><span></span><code>champagne<span class=w> </span>run<span class=w> </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/champagne_project<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--mode<span class=w> </span>slurm<span class=w> </span><span class=se>\</span>
<span class=w>    </span>-entry<span class=w> </span>MAKE_REFERENCE<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--genome<span class=w> </span>custom_genome<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--genome_fasta<span class=w> </span>genome.fasta<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--genes_gtf<span class=w> </span>genome.gtf<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--blacklist<span class=w> </span>blacklist.fasta
</code></pre></div> <p>The reference files and a config file for the genome will be written in <code>results/genome/custom_genome/</code>.</p> <p>Then you can run champagne using your custom genome:</p> <div class=highlight><pre><span></span><code>champagne<span class=w> </span>run<span class=w> </span>--output<span class=w> </span>/data/<span class=nv>$USER</span>/champagne_project<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--mode<span class=w> </span>slurm<span class=w> </span>-profile<span class=w> </span>biowulf<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--input<span class=w> </span>samplesheet.csv<span class=w> </span><span class=se>\</span>
<span class=w>    </span>--genome<span class=w> </span>custom_genome<span class=w> </span><span class=se>\</span>
<span class=w>    </span>-c<span class=w> </span>results/genome/custom_genome/custom_genome.config
</code></pre></div> </div> </div><script async src="https://www.googletagmanager.com/gtag/js?id=G-D3SL9V30KL"></script> <script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-D3SL9V30KL'); // Replace with your tracking ID
</script> </div> </div> </section> </div> <div class=rst-versions role=note aria-label=Versions> <span class=rst-current-version data-toggle=rst-current-version> <span> <a href=https://github.com/CCBR/CHAMPAGNE class="fa fa-code-fork" style="color: #fcfcfc"> CCBR/CHAMPAGNE</a> </span> <span><a href=../params/ style="color: #fcfcfc">&laquo; Previous</a></span> <span><a href=../spike-in/ style="color: #fcfcfc">Next &raquo;</a></span> </span> </div> <script src=../../js/jquery-3.6.0.min.js></script> <script>var base_url = "../..";</script> <script src=../../js/theme_extra.js></script> <script src=../../js/theme.js></script> <script src=../../js/custom.js></script> <script src=../../search/main.js></script> <script src=../../js/version-select.js></script> <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script> </body> </html> 