<!DOCTYPE html><html class=writer-html5 lang=en> <head><meta charset=utf-8><meta http-equiv=X-UA-Compatible content="IE=edge"><meta name=viewport content="width=device-width, initial-scale=1.0"><meta name=author content="CCR Collaborative Bioinformatics Resource"><link rel="shortcut icon" href=../../img/favicon.ico><title>Output - CHAMPAGNE</title><link rel=stylesheet href=../../css/theme.css><link rel=stylesheet href=../../css/theme_extra.css><link rel=stylesheet href=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css><link href=../../css/custom.css rel=stylesheet><link href=../../css/version-select.css rel=stylesheet><script>
        // Current page data
        var mkdocs_page_name = "Output";
        var mkdocs_page_input_path = "guide/output.md";
        var mkdocs_page_url = null;
      </script><!--[if lt IE 9]>
      <script src="../../js/html5shiv.min.js"></script>
    <![endif]--><script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js></script><script src=https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/languages/yaml.min.js></script><script>hljs.highlightAll();</script></head> <body class=wy-body-for-nav role=document> <div class=wy-grid-for-nav> <nav data-toggle=wy-nav-shift class="wy-nav-side stickynav"> <div class=wy-side-scroll> <div class=wy-side-nav-search> <a href=../.. class="icon icon-home"> CHAMPAGNE </a><div role=search> <form id=rtd-search-form class=wy-form action=../../search.html method=get> <input type=text name=q placeholder="Search docs" aria-label="Search docs" title="Type search term here"> </form> </div> </div> <div class="wy-menu wy-menu-vertical" data-spy=affix role=navigation aria-label="Navigation menu"> <ul> <li class=toctree-l1><a class="reference internal" href=../..>Home</a> </li> </ul> <p class=caption><span class=caption-text>User Guide</span></p> <ul class=current> <li class=toctree-l1><a href=../usage/ class="reference internal">Usage</a> </li> <li class="toctree-l1 current"><a class="reference internal current" href=#>Output</a> <ul class=current> <li class=toctree-l2><a class="reference internal" href=#results>Results</a> <ul> <li class=toctree-l3><a class="reference internal" href=#pipeline-information>Pipeline information</a> </li> <li class=toctree-l3><a class="reference internal" href=#quality-control>Quality Control</a> <ul> <li class=toctree-l4><a class="reference internal" href=#multiqc>MultiQC</a> </li> <li class=toctree-l4><a class="reference internal" href=#fastqc>fastQC</a> </li> <li class=toctree-l4><a class="reference internal" href=#deeptools>deepTools</a> </li> <li class=toctree-l4><a class="reference internal" href=#phantom-peak-qual-tools-ppqt>Phantom Peak Qual Tools (PPQT)</a> </li> </ul> </li> <li class=toctree-l3><a class="reference internal" href=#alignments>Alignments</a> </li> <li class=toctree-l3><a class="reference internal" href=#bigwigs>Bigwigs</a> </li> <li class=toctree-l3><a class="reference internal" href=#peaks>Peaks</a> <ul> <li class=toctree-l4><a class="reference internal" href=#differential-analysis>Differential analysis</a> </li> <li class=toctree-l4><a class="reference internal" href=#consensus-peaks>Consensus Peaks</a> </li> </ul> </li> <li class=toctree-l3><a class="reference internal" href=#custom-genome>Custom Genome</a> </li> </ul> </li> <li class=toctree-l2><a class="reference internal" href=#working-directory>Working Directory</a> </li> </ul> </li> <li class=toctree-l1><a href=../params/ class="reference internal">Parameters</a> </li> <li class=toctree-l1><a href=../genomes/ class="reference internal">Genomes</a> </li> <li class=toctree-l1><a href=../spike-in/ class="reference internal">Spike-in Normalization</a> </li> <li class=toctree-l1><a href=../troubleshooting/ class="reference internal">Troubleshooting</a> </li> </ul> <p class=caption><span class=caption-text>Developer Guide</span></p> <ul> <li class=toctree-l1><a href=../../devs/contributing/ class="reference internal">How to contribute</a> </li> <li class=toctree-l1><a href=../../devs/contributors/ class="reference internal">Contributors</a> </li> </ul> <ul> <li class=toctree-l1><a href=../../changelog/ class="reference internal">Changelog</a> </li> </ul> </div> </div> </nav> <section data-toggle=wy-nav-shift class=wy-nav-content-wrap> <nav class=wy-nav-top role=navigation aria-label="Mobile navigation menu"> <i data-toggle=wy-nav-top class="fa fa-bars"></i> <a href=../..>CHAMPAGNE</a> </nav> <div class=wy-nav-content> <div class=rst-content><div role=navigation aria-label="breadcrumbs navigation"> <ul class=wy-breadcrumbs> <li><a href=../.. class="icon icon-home" aria-label=Docs></a></li> <li class=breadcrumb-item>User Guide</li> <li class="breadcrumb-item active">Output</li> <li class=wy-breadcrumbs-aside> <a href=https://github.com/CCBR/CHAMPAGNE/edit/main/guide/output.md>Edit on CCBR/CHAMPAGNE</a> </li> </ul> <hr> </div> <div role=main class=document itemscope=itemscope itemtype=http://schema.org/Article> <div class=section itemprop=articleBody> <h1 id=output>Output<a class=headerlink href=#output title="Permanent link">&para;</a></h1> <p>When you initialize champagne, the project output directory (specified with <code>--output</code>) is populated with <code>assets/</code>, <code>conf/</code>, and <code>nextflow.config</code>. After running the pipeline, the output directory will also contain <code>results/</code>, <code>log/</code>, <code>work/</code>, and <code>submit_slurm.sh</code> (if using <code>--mode slurm</code>).</p> <div class=highlight><pre><span></span><code>/data/$USER/champagne_project
├── assets/
├── conf/
├── log/
├── nextflow.config
├── results/
├── submit_slurm.sh
└── work/
</code></pre></div> <table> <thead> <tr> <th>Directory</th> <th>Description</th> </tr> </thead> <tbody> <tr> <td><code>assets/</code></td> <td>Contains example sample sheets, contrasts, and other input files.</td> </tr> <tr> <td><code>conf/</code></td> <td>Contains configuration files for the pipeline.</td> </tr> <tr> <td><code>log/</code></td> <td>Contains log files for the pipeline run, including a slurm log if <code>--mode slurm</code> is used.</td> </tr> <tr> <td><code>nextflow.config</code></td> <td>The Nextflow configuration file for the pipeline.</td> </tr> <tr> <td><a href=#results><code>results/</code></a></td> <td>Contains the results files from the pipeline run.</td> </tr> <tr> <td><code>submit_slurm.sh</code></td> <td>A script to submit the pipeline run to SLURM (only created if <code>--mode slurm</code> is used).</td> </tr> <tr> <td><a href=#working-directory><code>work/</code></a></td> <td>Contains the working directory for Nextflow.</td> </tr> </tbody> </table> <h2 id=results>Results<a class=headerlink href=#results title="Permanent link">&para;</a></h2> <p>Workflow output files will be written to the <code>results/</code> directory by default in your pipeline run output directory.</p> <p>For example, if you ran champagne with <code>champagne run --output /data/$USER/champagne_project</code>, the results files will be in <code>/data/$USER/champagne_project/results/</code>.</p> <p>All paths listed below are relative to the <code>results/</code> directory.</p> <h3 id=pipeline-information>Pipeline information<a class=headerlink href=#pipeline-information title="Permanent link">&para;</a></h3> <p><code>pipeline_info/</code> contains information about the pipeline run.</p> <p><code>pipeline_info/bco.json</code> is the <a href=https://biocomputeobject.org/ >BioCompute Object (BCO)</a> for the pipeline run, which is a standardized format for sharing computational workflows and results.</p> <p><code>pipeline_info/dag.html</code> is the DAG (Directed Acyclic Graph) of the pipeline run, which shows the workflow execution order and dependencies.</p> <p>The <a href=https://www.nextflow.io/docs/latest/reports.html#execution-report>execution report</a>, <a href=https://www.nextflow.io/docs/latest/reports.html#execution-timeline>timeline</a>, and <a href=https://www.nextflow.io/docs/latest/reports.html#trace-file>trace file</a> generated by Nextflow are also included in this directory and are named with the timestamp of the pipeline run.</p> <h3 id=quality-control>Quality Control<a class=headerlink href=#quality-control title="Permanent link">&para;</a></h3> <p><code>qc/</code> contains all quality control files.</p> <h4 id=multiqc>MultiQC<a class=headerlink href=#multiqc title="Permanent link">&para;</a></h4> <p><code>qc/multiqc/multiqc_report.html</code> is the <a href=https://seqera.io/multiqc/ >MultiQC</a> report for the entire run, which summarizes the quality control results for all samples.</p> <p>Input files passed to multiqc are in <code>qc/multiqc/input/</code>.</p> <h4 id=fastqc>fastQC<a class=headerlink href=#fastqc title="Permanent link">&para;</a></h4> <p><code>qc/fastqc/raw</code> and <code>qc/fastqc/trimmed</code> contain FastQC reports for raw and trimmed reads, respectively.</p> <h4 id=deeptools>deepTools<a class=headerlink href=#deeptools title="Permanent link">&para;</a></h4> <p><code>qc/deeptools/</code> contains plots and intermediate files generated by deepTools, including QC metrics, PCA plots, scatter plots, and correlation heatmaps.</p> <h4 id=phantom-peak-qual-tools-ppqt>Phantom Peak Qual Tools (PPQT)<a class=headerlink href=#phantom-peak-qual-tools-ppqt title="Permanent link">&para;</a></h4> <p><code>qc/phantompeakqualtools/</code> contains the PPQT reports and fragment lengths for each sample.</p> <h3 id=alignments>Alignments<a class=headerlink href=#alignments title="Permanent link">&para;</a></h3> <p><code>align/bam/</code> contains the sorted and deduped BAM files for each sample.</p> <h3 id=bigwigs>Bigwigs<a class=headerlink href=#bigwigs title="Permanent link">&para;</a></h3> <p><code>bigwigs/</code> contains the bigwig files for each sample; one before removing input reads (<code>{id}.bw</code>) and one with input reads removed (<code>{id}.inputnorm.bw</code>).</p> <p>These are normalized with the method set by the <code>deeptools_normalize_using</code> parameter.</p> <p>If using a spike-in genome, these are normalized according the spike-in parameters set. See the <a href=../spike-in/ >Spike-in normalization doc</a> for more details.</p> <h3 id=peaks>Peaks<a class=headerlink href=#peaks title="Permanent link">&para;</a></h3> <p><code>peaks/{peak_caller}/</code> directories are created for each peak caller. Available peak callers are <code>macs2_narrow</code>, <code>macs2_broad</code>, <code>gem</code>, and <code>sicer</code>. See the <a href=../params/#run-control>run parameters</a> and <a href=../params/#peak-caller>peak caller parameters</a> for customization options.</p> <p><code>peaks/{peak_caller}/replicates/</code> contains called peaks for each sample.</p> <h4 id=differential-analysis>Differential analysis<a class=headerlink href=#differential-analysis title="Permanent link">&para;</a></h4> <p>If any sample has only one replicate, manorm is used for differential analysis. Otherwise, diffbind is used.</p> <p>manorm results are in <code>peaks/{peak_caller}/diff/manorm/{contrast}/</code>.</p> <p>diffbind results are in <code>peaks/{peak_caller}/diff/diffbind/{contrast}/</code>.</p> <h4 id=consensus-peaks>Consensus Peaks<a class=headerlink href=#consensus-peaks title="Permanent link">&para;</a></h4> <p><code>peaks/{peak_caller}/consensus/{consensus_method}</code> contains the consensus peaks for each peak caller.</p> <h5 id=motifs>Motifs<a class=headerlink href=#motifs title="Permanent link">&para;</a></h5> <p><code>peaks/{peak_caller}/consensus/{consensus_method}/motifs</code> contains HOMER and MEME AME results.</p> <p>Homer results are in <code>motifs/homer/</code>, organized with one directory per sample (e.g. <code>{id}_homer/</code>) and contain a background fasta, target fasta, results HTML, and a subdirectory containing motif files and logos.</p> <p>If the target and background fasta files from homer are not blank, they are used for MEME AME for motif enrichment analysis. These results files are in <code>motifs/meme/</code>.</p> <h5 id=annotations>Annotations<a class=headerlink href=#annotations title="Permanent link">&para;</a></h5> <p><code>peaks/{peak_caller}/consensus/{consensus_method}/annotations</code> contains peak annotations from ChIPseeker.</p> <h3 id=custom-genome>Custom Genome<a class=headerlink href=#custom-genome title="Permanent link">&para;</a></h3> <p>If you specified a custom genome with <a href=../params/#custom-genome-options>custom genome parameters</a>, the prepared genome files will be in <code>genome/{genome}</code>. These include the genome BWA index, blacklist BWA index, and a nextflow config file to reuse the genome in future runs.</p> <h2 id=working-directory>Working Directory<a class=headerlink href=#working-directory title="Permanent link">&para;</a></h2> <p><code>work/</code> is where Nextflow stores intermediate files during the pipeline run. If you resubmit the pipeline, the working directory will be reused so the pipeline can resume from where it previously stopped. After successfully completing the pipeline run, you can delete the working directory to save disk space.</p> </div> </div><script async src="https://www.googletagmanager.com/gtag/js?id=G-D3SL9V30KL"></script> <script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());
  gtag('config', 'G-D3SL9V30KL'); // Replace with your tracking ID
</script> </div> </div> </section> </div> <div class=rst-versions role=note aria-label=Versions> <span class=rst-current-version data-toggle=rst-current-version> <span> <a href=https://github.com/CCBR/CHAMPAGNE class="fa fa-code-fork" style="color: #fcfcfc"> CCBR/CHAMPAGNE</a> </span> <span><a href=../usage/ style="color: #fcfcfc">&laquo; Previous</a></span> <span><a href=../params/ style="color: #fcfcfc">Next &raquo;</a></span> </span> </div> <script src=../../js/jquery-3.6.0.min.js></script> <script>var base_url = "../..";</script> <script src=../../js/theme_extra.js></script> <script src=../../js/theme.js></script> <script src=../../js/custom.js></script> <script src=../../search/main.js></script> <script src=../../js/version-select.js></script> <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script> </body> </html> 